File: projecteuler/24.py
# A permutation is an ordered arrangement of objects. For example, 3124 is one
#  possible permutation of the digits 1, 2, 3 and 4. If all of the permutations
#  are listed numerically or alphabetically, we call it lexicographic order.
# The lexicographic permutations of 0, 1 and 2 are:
#	012   021   102   120   201   210
# What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9?


# Add factorials while less than 1 million, using division to find digit

def fact(n):
	if n == 1 or n == 0:
		return 1
	return n * fact(n-1)

N = 1000001

total = N
digits = "12345679"
for i in range(10):
	print(str(i) in digits)

File: projecteuler/24.rs
/*
Problem 24: Find millionth lexicographic permutation of the decimal digits (0-9).

Total permutations is 10!
The first 9! permutations begin with 0, the second 9! begin with 1, etc.
Use this logic to iteratively subtract the k'th factorial till one mil is reached.

Can easily verify that the millionth permutation will start with '2' (the third
 digit), since 2*9! < 1000000 < 3*9!
So the remaining digits are (0,1,3,4,5,6,7,8,9)
*/

fn factorial(n: usize) -> usize {
	let mut ret = 1;
	for i in 2..=n {
		ret *= i;
	}

	ret
}

fn main() {
	const N: usize = 1000000;

	// 0-indexed remainder
	let mut remainder: usize = N-1;

	let mut digits: Vec<usize> = (0..=9).collect();
	println!("digits = {digits:?}");

	let mut k = digits.len() - 1;
	let mut result = Vec::new();
	while remainder > 0 {
		let pos = remainder / factorial(k);

		let d = digits.remove(pos);
		result.push(d);

		remainder -= pos * factorial(k);
		k -= 1;
		println!("Remainder: {remainder}");
	}

	// Append remaining digits (front-to-back, increasing order)
	result.extend_from_slice(&digits);

	let s: String = result.into_iter().map(|x| x.to_string()).collect();
	println!("Result = {}", s);
} 
